toss/fitness/fitness_function_utils.py

```diff
@@ -45,4 +45,68 @@ def _compute_squared_distance(positions,constant):
+    and subsequently reassign the visited boolean value to true. The ratio of visited points 
+    is then given by the number of True values to the total number of values in the boolean array.
+
+    Args:
+        positions (np.ndarray): (3,N) Array of positions (in cartesian coordinates).
+        velocities (np.ndarray): (3,N) Array of positions (in cartesian frame).
+        timesteps (np.ndarray): (N) Array of time values for each position.
+        radius_outer_bounding_sphere (float): Radius of outer bounding sphere.
+
+    Returns:
+        ratio (float): Number of True values to the total number of values in the boolean array
+    """
+
+    # Define frequency of points for the spherical meshgrid: (see: Courant–Friedrichs–Lewy condition)
+    cfl = 1 # Courant number
+    v_max = np.sqrt(np.max(velocities[0,:]**2 + velocities[1,:]**2 + velocities[2,:]**2))
+    dt = timesteps[1]-timesteps[0]
+    
+    # Calculate and adjust grid spacing based on the Courant number, maximal velocity, and time step
+    grid_spacing = v_max * dt / cfl
+    r = np.linspace(0, radius_outer_bounding_sphere, int(radius_outer_bounding_sphere / grid_spacing)) # Number of evenly spaced points along the radial axis
+    theta = np.linspace(0, np.pi, int(np.pi / grid_spacing)) # Number of evenly spaced points along the polar angle (defined in between 0 and pi)
+    phi = np.linspace(0, 2 * np.pi, int(2 * np.pi / grid_spacing)) # Number of evenly spaced points along the azimuthal angle (defined in between 0 and 2*pi)
+
+    # Create a spherical meshgrid
+    r_matrix, theta_matrix, phi_matrix = np.meshgrid(r, theta, phi)
+
+    # Convert the positions along the trajectory to spherical coordinates
+    r_points = np.sqrt(positions[0,:]**2 + positions[1,:]**2 + positions[2,:]**2) # radial coordinates of points
+    theta_points = np.arccos(positions[2,:] / r_points) # polar angle coordinates of points
+    phi_points = np.arctan2(positions[1,:], positions[0,:]) # azimuthal angle coordinates of points
+
+    # Remove points outside measurement zone (i.e outside outer-bounding sphere)
+    index_feasible_positions = r_points < radius_outer_bounding_sphere
+    r_points = r_points[index_feasible_positions]
+    theta_points = theta_points[index_feasible_positions]
+    phi_points = phi_points[index_feasible_positions]
+
+    # Find the indices of the closest values in the meshgrid for each point using broadcasting
+    i = np.argmin(np.abs(r[:, np.newaxis] - r_points), axis=0) # indices along r axis
+    j = np.argmin(np.abs(theta[:, np.newaxis] - theta_points), axis=0) # indices along theta axis
+    k = np.argmin(np.abs(phi[:, np.newaxis] - phi_points), axis=0) # indices along phi axis
+
+    # Create a boolean tensor with the same shape as the spherical meshgrid
+    bool_array = np.zeros_like(r_matrix, dtype=bool) # initialize with False
+
+    # Set the values to True where the points are located using advanced indexing
+    bool_array[j, i, k] = True
```

Comment:
why j,i.k, shoudln't it be i,j,k as before?

this might be a good opportunity to write a very quick test where you either randomly sample a million points and just check it doesn't crash for them or systematically go over the points in the mesh grid and check that "visiting" them leads to a ratio of 1.0 and no crashes?

Reply:
It is actually correct. np.meshrgid returns the r_matrix with dimensions according to j,i,k for some reason. I've checked this by comparing withe the length of the arrays for r, theta and phi.